-1

**Failed Tests**:   Build  ClangBuild 
**Size**: This PR adds an extra 20KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-0a3639/50821/summary.html
**COMMIT**: e6e7188d4fac9d7c83ce321617a7b60b7631bfb1
**CMSSW**: CMSSW_16_1_X_2026-01-22-1100/el8_amd64_gcc13
**Additional Tests**: GPU,AMD_MI300X,AMD_W7900,NVIDIA_H100,NVIDIA_L40S,NVIDIA_T4
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/49856/50821/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Failed Build

I found compilation error when building: 

<pre>>> Compiling alpaka/cuda src/HeterogeneousCore/AlpakaCore/src/alpaka/EDMetadataSentry.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/gcc/13.4.0-6908cfdf803923e783448096ca4f0923/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=130400 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DBOOST_MPL_IGNORE_PARENTHESES_WARNING -DCMSSW_GIT_HASH='CMSSW_16_1_X_2026-01-22-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_16_1_X_2026-01-22-1100' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/cms/cmssw-patch/CMSSW_16_1_X_2026-01-22-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/pcre/8.43-6d98fda3bfd074ebb583e2d6a2c75d25/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/boost/1.80.0-b819d3899535842b3b08dcd6a725af1a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/bz2lib/1.0.6-d113e1c6278c07eeaff5f84db9548446/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/libuuid/2.34-5ba7a8abfc0c5fecdc448cca360c25ff/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/lcg/root/6.36.07-bd71f88d9cd20e5042c0ac03a7e23595/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/tbb/v2022.3.0-88eb7be4ee320d604a798a914aea6359/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/xz/5.6.4-b9c4ffbc390ed320a5d57fd552e29a05/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/zlib/1.2.13-589f6bb51bbeba38a7adf5a10ea8a093/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/alpaka/2.1.1-3caaac8d71f39d400ab2511b2403675a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/cuda/12.9.1-cff83d5f72da96ebfea8cafd87a05296/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/fmt/10.2.1-31d67b0504b4ba2262f03d3c5cad83c1/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/md5/1.0.0-26057075013e190e56dad37d35219376/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/tinyxml2/6.2.0-67924ead96ecb4e69aad321b767979a5/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DALPAKA_HAS_STD_ATOMIC_REF -Wno-error=unused-variable -DALPAKA_ACC_GPU_CUDA_ENABLED -DALPAKA_ACC_GPU_CUDA_ONLY_MODE -DALPAKA_HOST_ONLY -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc13/src/HeterogeneousCore/AlpakaCore/src/alpaka/HeterogeneousCoreAlpakaCoreCudaAsync/EDMetadataSentry.cc.d src/HeterogeneousCore/AlpakaCore/src/alpaka/EDMetadataSentry.cc -o tmp/el8_amd64_gcc13/src/HeterogeneousCore/AlpakaCore/src/alpaka/HeterogeneousCoreAlpakaCoreCudaAsync/EDMetadataSentry.cc.o
>> Compiling alpaka/cuda src/HeterogeneousCore/AlpakaCore/src/alpaka/ESProducer.cc
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/gcc/13.4.0-6908cfdf803923e783448096ca4f0923/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=130400 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DBOOST_MPL_IGNORE_PARENTHESES_WARNING -DCMSSW_GIT_HASH='CMSSW_16_1_X_2026-01-22-1100' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_16_1_X_2026-01-22-1100' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/cms/cmssw-patch/CMSSW_16_1_X_2026-01-22-1100/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/pcre/8.43-6d98fda3bfd074ebb583e2d6a2c75d25/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/boost/1.80.0-b819d3899535842b3b08dcd6a725af1a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/bz2lib/1.0.6-d113e1c6278c07eeaff5f84db9548446/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/libuuid/2.34-5ba7a8abfc0c5fecdc448cca360c25ff/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/lcg/root/6.36.07-bd71f88d9cd20e5042c0ac03a7e23595/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/tbb/v2022.3.0-88eb7be4ee320d604a798a914aea6359/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/xz/5.6.4-b9c4ffbc390ed320a5d57fd552e29a05/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/zlib/1.2.13-589f6bb51bbeba38a7adf5a10ea8a093/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/alpaka/2.1.1-3caaac8d71f39d400ab2511b2403675a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/cuda/12.9.1-cff83d5f72da96ebfea8cafd87a05296/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/fmt/10.2.1-31d67b0504b4ba2262f03d3c5cad83c1/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/md5/1.0.0-26057075013e190e56dad37d35219376/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02925/el8_amd64_gcc13/external/tinyxml2/6.2.0-67924ead96ecb4e69aad321b767979a5/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -DALPAKA_DEFAULT_HOST_MEMORY_ALIGNMENT=128 -DALPAKA_DISABLE_VENDOR_RNG -DALPAKA_HAS_STD_ATOMIC_REF -Wno-error=unused-variable -DALPAKA_ACC_GPU_CUDA_ENABLED -DALPAKA_ACC_GPU_CUDA_ONLY_MODE -DALPAKA_HOST_ONLY -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc13/src/HeterogeneousCore/AlpakaCore/src/alpaka/HeterogeneousCoreAlpakaCoreCudaAsync/ESProducer.cc.d src/HeterogeneousCore/AlpakaCore/src/alpaka/ESProducer.cc -o tmp/el8_amd64_gcc13/src/HeterogeneousCore/AlpakaCore/src/alpaka/HeterogeneousCoreAlpakaCoreCudaAsync/ESProducer.cc.o
src/HeterogeneousCore/AlpakaCore/src/alpaka/EDMetadata.cc: In member function 'virtual void alpaka_cuda_async::EDMetadata::synchronize() const':
src/HeterogeneousCore/AlpakaCore/src/alpaka/EDMetadata.cc:48:7: error: expected primary-expression before 'catch'
   48 |       catch (...) {
      |       ^~~~~
src/HeterogeneousCore/AlpakaCore/src/alpaka/EDMetadata.cc:51:5: error: expected 'catch' before 'eventComplete_'
   51 |     eventComplete_ = true;
      |     ^~~~~~~~~~~~~~
</pre>

## Failed Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Local Products Rules ..... done
>> Creating project symlinks
>> Entering Package DataFormats/Common
>> Entering Package HeterogeneousCore/AlpakaCore
>> Compile sequence completed for CMSSW CMSSW_16_1_X_2026-01-22-1100
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 1
Command exited with non-zero status 1
	Command being timed: "scram build -k -j 32 COMPILER=llvm compile BUILD_LOG=yes"
	User time (seconds): 125.28
	System time (seconds): 17.64
	Percent of CPU this job got: 1057%
</pre>

--- a/HeterogeneousCore/AlpakaCore/interface/alpaka/EDMetadata.h
+++ b/HeterogeneousCore/AlpakaCore/interface/alpaka/EDMetadata.h
@@ -6,6 +6,7 @@
 
 #include <alpaka/alpaka.hpp>
 
+#include "DataFormats/Common/interface/DeviceProduct.h"
 #include "FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h"
 #include "HeterogeneousCore/AlpakaInterface/interface/config.h"
 
@@ -62,11 +63,11 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
 #else
   // All backends with an asynchronous queue
 
-  class EDMetadata {
+  class EDMetadata : public edm::DeviceProductMetadataBase {
   public:
     EDMetadata(std::shared_ptr<Queue> queue, std::shared_ptr<Event> event)
         : queue_(std::move(queue)), event_(std::move(event)) {}
-    ~EDMetadata();
+    ~EDMetadata() noexcept override;
 
     Device device() const { return alpaka::getDev(*queue_); }
 
@@ -84,7 +85,16 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
     void discardEvent() { event_.reset(); }
 
     /**
+     * Synchronizes 'this' metadata wrt. host (caller)
+     * This is always a blocking call. It is intended to be called from
+     * edm::DeviceProduct destructor.
+     */
+    void synchronize() const noexcept final;
+
+    /**
      * Synchronizes 'consumer' metadata wrt. 'this' in the event product
+     * This is a non-blocking call. At the worst case the queue of the 'consumer'
+     * will wait for the event of 'this' to complete.
      */
     void synchronize(EDMetadata& consumer, bool tryReuseQueue) const;
 

--- a/HeterogeneousCore/AlpakaCore/src/alpaka/EDMetadata.cc
+++ b/HeterogeneousCore/AlpakaCore/src/alpaka/EDMetadata.cc
@@ -7,29 +7,7 @@
 
 namespace ALPAKA_ACCELERATOR_NAMESPACE {
 #ifndef ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED
-  EDMetadata::~EDMetadata() {
-    // Make sure that the production of the product in the GPU is
-    // complete before destructing the product. This is to make sure
-    // that the EDM stream does not move to the next event before all
-    // asynchronous processing of the current is complete.
-
-    // TODO: a callback notifying a WaitingTaskHolder (or similar)
-    // would avoid blocking the CPU, but would also require more work.
-
-    // If event_ is null, the EDMetadata was either
-    // default-constructed, or fully synchronized before leaving the
-    // produce() call, so no synchronization is needed.
-    // If the queue was re-used, then some other EDMetadata object in
-    // the same edm::Event records the event_ (in the same queue) and
-    // calls alpaka::wait(), and therefore this wait() call can be
-    // skipped).
-    if (event_ and not eventComplete_ and mayReuseQueue_) {
-      // Must not throw in a destructor, and if there were an
-      // exception could not really propagate it anyway.
-      CMS_SA_ALLOW try { alpaka::wait(*event_); } catch (...) {
-      }
-    }
-  }
+  EDMetadata::~EDMetadata() noexcept { synchronize(); }
 
   void EDMetadata::enqueueCallback(edm::WaitingTaskWithArenaHolder holder) {
     edm::Service<edm::Async> async;
@@ -38,6 +16,34 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE {
         std::move(holder),
         [event = event_]() mutable { alpaka::wait(*event); },
         []() { return "Enqueued via " EDM_STRINGIZE(ALPAKA_ACCELERATOR_NAMESPACE) "::EDMetadata::enqueueCallback()"; });
+  }
+
+  void EDMetadata::synchronize() const noexcept {
+    // Make sure that the production of the product in the GPU is
+    // complete before destructing the product. This will also make sure
+    // that the EDM stream does not move to the next event before all
+    // asynchronous processing of the current is complete.
+
+    // TODO: a callback notifying a WaitingTaskHolder (or similar)
+    // would avoid blocking the CPU, but would also require more work.
+
+    if (eventComplete_) {
+      return;
+    }
+
+    // If the event has been discarded, the produce() function that
+    // constructed this EDMetadata object did not launch any
+    // asynchronous work.
+    if (not event_) {
+      return;
+    }
+
+    // This function is called from a destructor, so we want
+    // to avoid throwing exceptions. If there was an
+    // exception we could not really propagate it anyway.
+    CMS_SA_ALLOW try { alpaka::wait(*event_); } catch (...) {
+    }
+    eventComplete_ = true;
   }
 
   void EDMetadata::synchronize(EDMetadata& consumer, bool tryReuseQueue) const {

